src/category_theory/idempotents/functor_categories.lean

```diff
@@ -0,0 +1,166 @@
+`karoubi_functor_category_embedding : karoubi (J ⥤ C)) : J ⥤ karoubi C` for all categories
+`J` and `C`.
+
+-/
+
+open category_theory
+open category_theory.category
+open category_theory.idempotents.karoubi
+open category_theory.limits
+
+namespace category_theory
+
+namespace idempotents
+
+variables (J C : Type*) [category J] [category C]
+
+instance functor_category_is_idempotent_complete [is_idempotent_complete C] :
+  is_idempotent_complete (J ⥤ C) :=
+begin
+  refine ⟨_⟩,
+  intros F p hp,
+  have hC := (is_idempotent_complete_iff_has_equalizer_of_id_and_idempotent C).mp infer_instance,
+  haveI : ∀ (j : J), has_equalizer (𝟙 _) (p.app j) := λ j, hC _ _ (congr_app hp j),
+  /- We construct the direct factor `Y` associated to `p : F ⟶ F` by computing
+    the equalizer of the identity and `p.app j` on each object `(j : J)`.  -/
+  let Y : J ⥤ C :=
+  { obj := λ j, limits.equalizer (𝟙 _) (p.app j),
+    map := λ j j' φ, equalizer.lift (limits.equalizer.ι (𝟙 _) (p.app j) ≫ F.map φ)
+      (by rw [comp_id, assoc, p.naturality φ, ← assoc, ← limits.equalizer.condition, comp_id]),
+    map_id' := λ j, by { ext, simp only [comp_id, functor.map_id, equalizer.lift_ι, id_comp], },
+    map_comp' := λ j j' j'' φ φ', begin
+      ext,
+      simp only [assoc, functor.map_comp, equalizer.lift_ι, equalizer.lift_ι_assoc],
+    end },
+  let i : Y ⟶ F :=
+  { app := λ j, equalizer.ι _ _,
+    naturality' := λ j j' φ, by rw [equalizer.lift_ι],  },
+  let e : F ⟶ Y :=
+  { app := λ j, equalizer.lift (p.app j)
+      (by { rw comp_id, exact (congr_app hp j).symm, }),
+    naturality' := λ j j' φ, begin
+      ext,
+      simp only [assoc, equalizer.lift_ι, nat_trans.naturality, equalizer.lift_ι_assoc],
+    end },
+  use [Y, i, e],
+  split; ext j,
+  { simp only [nat_trans.comp_app, assoc, equalizer.lift_ι, nat_trans.id_app, id_comp,
+      ← equalizer.condition, comp_id], },
+  { simp only [nat_trans.comp_app, equalizer.lift_ι], },
+end
+
+namespace karoubi_functor_category_embedding
+
+variables {J C}
+
+/-- On objects, the functor which sends a formal direct factor `P` of a
+functor `F : J ⥤ C` to the functor `J ⥤ karoubi C` which sends `(j : J)` to
+the corresponding direct factor of `F.obj j`. -/
+@[simps]
+def obj (P : karoubi (J ⥤ C)) : J ⥤ karoubi C :=
+{ obj := λ j, ⟨P.X.obj j, P.p.app j, congr_app P.idempotence j⟩,
+  map := λ j j' φ,
+  { f := P.p.app j ≫ P.X.map φ,
+    comm := begin
+      simp only [nat_trans.naturality, assoc],
+      have h := congr_app P.idempotence j,
+      rw [nat_trans.comp_app] at h,
+      slice_rhs 1 3 { erw [h, h], },
+    end },
+  map_id' := λ j, by { ext, simp only [functor.map_id, comp_id, id_eq], },
+  map_comp' := λ j j' j'' φ φ', begin
+    ext,
+    have h := congr_app P.idempotence j,
+    rw [nat_trans.comp_app] at h,
+    simp only [assoc, nat_trans.naturality_assoc, functor.map_comp, comp],
+    slice_rhs 1 2 { rw h, },
+    rw [assoc],
+  end }
+
+/-- Action of maps on `obj`. -/
```

Comment:
This docstring is not very helpful outside the context of this file. Please extend it.

Reply:
Thanks for the suggestion.